noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays/assert_split_transformed_padded_arrays.nr

```diff
@@ -0,0 +1,404 @@
+    num_padded_lt: u32,
+    first_after_split_index: u32,
+    first_padded_index: u32,
+)
+where
+    T: Ordered + Empty + Eq,
+    S: Empty + Eq,
+{
+    // Validate the hints
+    //
+    // Validate first_after_split_index
+    if first_after_split_index != 0 {
+        assert(
+            sorted_array[first_after_split_index - 1].counter() < split_counter,
+            "first_after_split_index is too large",
+        );
+    }
+    if first_after_split_index != N {
+        let counter = sorted_array[first_after_split_index].counter();
+        // Note: A padded item can also be the first item after the split counter.
+        assert((counter == 0) | (counter >= split_counter), "incorrect first_after_split_index");
+    }
+    // Validate first_padded_index
+    if first_padded_index != 0 {
+        assert(
+            sorted_array[first_padded_index - 1].counter() != MAX_U32_VALUE,
+            "first_padded_index is too large",
+        );
+    }
+    if first_padded_index != N {
+        assert(
+            sorted_array[first_padded_index].counter() == MAX_U32_VALUE,
+            "incorrect first_padded_index",
+        );
+    }
+    // Validate num_padded_lt/first_padded_gte_index
+    let first_padded_gte_index = first_padded_index + num_padded_lt;
+    if first_padded_gte_index != N {
+        // If num_padded_lt is too large and there isn't enough padded items,
+        // first_padded_gte_index will be > N and the following will fail because the index is out of bounds.
+        let counter = sorted_array[first_padded_gte_index].counter();
+        // If there's no padded items for gte, the index will point to an empty item. And we can treat them as regular
+        // items and "pad" them to the gte array.
+        assert((counter == 0) | (counter == MAX_U32_VALUE), "incorrect first_padded_gte_index");
+    }
+
+    let mut is_lt = true;
+    let mut is_non_padded_lt = true;
+    let mut is_padded = false;
+    let mut is_padded_gte = false;
+    let num_non_padded_lt = first_after_split_index;
+    let num_non_padded_gte = first_padded_index - first_after_split_index;
+    let num_lt = num_non_padded_lt + num_padded_lt;
+    for i in 0..N {
+        is_non_padded_lt &= i != first_after_split_index;
+        is_padded |= i == first_padded_index;
+        is_padded_gte |= i == first_padded_gte_index;
+        is_lt = is_non_padded_lt | (is_padded & !is_padded_gte);
+
+        let from = sorted_array[i];
+        let to_index = if is_non_padded_lt {
+            i
+        } else if is_padded_gte {
+            i - num_lt
+        } else if is_padded {
+            i - num_non_padded_gte
+        } else {
+            i - num_non_padded_lt
+        };
+        let to = if is_lt {
+            transformed_value_array_lt[to_index]
+        } else {
+            transformed_value_array_gte[to_index]
+        };
+        assert(is_transformed(from, to), "invalid transformed value");
```

Comment:
I found this a bit hard to follow at first, the naming is good and clear but it was difficult to keep in mind all the possible cases! I'm not sure if it can be simplified, but maybe a separate method like `get_target_index` could be used like `let to_index = get_target_index(i, flags)` just to break it up a little? That also may be more confusing though so happy to discuss